docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,450 @@
+
+### Customizing A Workload By Region
+
+### Customizing A Workload By Cluster and Region
+
+### Customizing A Workload By Environment
+
+## Detailed Design
+
+### `PackageVariant` API
+
+#### Basic API
+
+- Upstream
+- Downstream
+- Annotations
+- Labels
+- Adoption and Deletion Policy
+
+#### Package Context Injection
+
+#### KRM Function Pipeline
+
+#### Configuration Injection
+
+Configuration injection is controlled by a combination of in-package resources
+with annotations, and *injection selectors* defined on the PackageVariant
+resource.
+
+Injection selectors are defined in the `spec.injectionSelectors` field of the
+`PackageVariant`. This field is an ordered array of structs containing group,
+version, kind, and name. Only the name is required.
+
+The injection process works as follows:
+
+1. The controller will examine all in-package resources, looking for those with
+   an annotation named `kpt.dev/config-injection`, with one of the following
+   values: `required` or `optional`. We will call these "injection points".
+1. For each injection point, a condition will be created in the
+   downstream PackageRevision, with ConditionType set to the dot-delimited
+   concatenation of `config.injection`, with the in-package resource kind and
+   name, and the value set to `False`. Note that since the package author
+   controls the name of the resource, kind and name are sufficient to
+   disambiguate the injection point. We will call this ConditionType the
+   "injection point ConditionType".
+1. For each required injection point, the injection point ConditionType will be
+   added to the PackageRevision `readinessGates`. Optional injection points'
+   ConditionTypes must not be added to the readinessGates by the PackageVariant
+   controller, but humans or other actors may do so at a later date, and the
+   PackageVariant controller should not remove them on subsequent
+   reconciliations. Also, this relies upon `readinessGates` preventing
+   publishing to a *deployment* repository, but should not blueprint
+   repositories.
+1. The injection processing will proceed as follows. For each injection point:
+   - If the resource schema of the injection point is not available in the
+     cluster, then the injection point ConditionType will be set to `False`,
+     with a message indicating that the schema is missing, and processing should
+     proceed to the next injection point. Note that for `optional` injection
+     points, not having the schema may be intentional and not an error.
+   - If the resource schema of the injection point does not contain a `spec`
+     field, then the injection point ConditionType will be set to `False`, with
+     a message explaining the error, and processing should proceed to the next
+     injection point.
+   - The controller will now identify all in-cluster objects in the same
+     namespace as the PackageVariant resource, with GVK matching the injection
+     point (the in-package resource).
```

Comment:
Right, the names will not match. For one, the names in the package are the same across for any PackageVariant using that upstream, but we want to inject different resources for each one. For example, in the package it may be named "cluster-info" but when we instantiate the PackageVariant for "cluster A", we want to inject the resource corresponding to "clusterA". We also do not want to *change* the name in the package, because it likely has meaning within the package and will be used by functions in the package.

Also, different roles that are controlling the names. The names in the package are in the control of the package author. The names in the cluster are in the control of whoever populates the cluster (could be, for example, some infrastructure team). The selector is the glue between them, and *that* is in control of the package author, and will be used in function configs, for example.

The GVK on the other hand, has to be the same, because it tells us the API. The namespace has to be the same, just for authorization purposes.

Reply:
I added a bunch of this explanation into the file.

Reply:
Thanks! The examples are really helpful